### DISCUSSION TITLE
`==`과 equals()의 차이

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#20] [Java]
.

**Comment by kunsanglee** [comment]
동등성과 동일성이 무엇인지 같이 질문하면 더 좋겠네요

**Comment by jimi567** [comment]
동일성(==)은 두 개의 변수가 메모리 상에서 동일한 객체를 참조하고 있는지를 확인한다. 객체의 내용과 관련없이 참조 주소를 비교하기에, 두 객체가 동일한 메모리 주소를 가리키는 경우에만 true를 반환한다.
그에 반해 동등성의 경우 객체의 내용이 동일한지를 비교한다. 즉 두 객체가 같은 값을 가지고 있는지를 확인한다.

equals 메서드는 기본적으로 Object 클래스에서 정의되어 있으며, 하위 클래스에서 필요에 따라 오버라이드 할 수 있다.
일반적으로 사용되는 equals 메서드는 객체의 내부 데이터나 상태를 비교하여, 그 결과에 따라 true 또는 false를 반환하게 된다.

**Comment by nayonsoso** [comment]
어떤 두 객체가 "같은 객체"인지 판단하기 위해서는
어떤 관계여야 '같다고 할 것인가?'에 대한 기준이 필요합니다.
동일성과 동등성은 여기에 대한 기준이 됩니다.

동일성은, 객체가 참조하는 주소가 같으면 같다고 판단합니다.
반면 동등성은, 개발자가 정의하는 기준에 따라서 같은 객체이면 같다고 판단합니다.

자바에서는 동일성을 비교하는데에 주로 `==`를,
동등성을 비교하는데에는 주로 `개발자가 재정의한 equals`를 사용합니다.
동등성 비교의 대표적인 예로는 String 비교가 있습니다.
